### PR TITLE
bindings for hpx/algorithm.hpp

### PR DESCRIPTION
**Pull request by pingu-73**
## Algos Done:
- [x] hpx::copy
- [x] hpx::copy_if
- [x] hpx::copy_n
- [x] hpx::count
- [x] hpx::count_if
- [x] hpx::ends_with
- [x] hpx::equal
- [x] hpx::fill (only works with linear vectors)
- [x] hpx::find
- [x] hpx::sort (with & without comparator)
- [x] hpx::merge
- [x] hpx::partial_sort (with and without comparator)

### CODE REVIEW

**Comment by pingu-73** [comment]
@SAtacker my hpx::copy_n is wrong, currently for first argument(as per docs refer to `beginning of the sequence of elements`) it would consider starting of src vec.
To start copying from some other location in the src vector should i:
- increase arguments and add a offset variable
- or use slice?

File: hpx-sys/src/lib.rs [hunk 1]
```diff
@@ -19,7 +18,11 @@ pub mod ffi {
         fn terminate();
         fn disconnect() -> i32;
         fn disconnect_with_timeout(shutdown_timeout: f64, localwait: f64) -> i32;
-        //fn stop() -> i32;
+        fn hpx_copy(src: &Vec<i32>, dest: &mut Vec<i32>);
+        fn hpx_copy_n(src: &Vec<i32>, count: usize, dest: &mut Vec<i32>);
+        fn hpx_copy_if(src: &Vec<i32>, dest: &mut Vec<i32>, pred: fn(i32) -> bool);
+        fn hpx_count(vec: &Vec<i32>, value: i32) -> i64;
+        fn hpx_count_if(vec: &Vec<i32>, pred: fn(i32) -> bool) -> i64;
```
Comment (SAtacker) [inline hpx-sys/src/lib.rs]: Is there any way it can be a generic vector type in rust?

Comment (pingu-73) [reply hpx-sys/src/lib.rs]: I can only do it this [way](https://codeberg.org/pingu-73/demo.git) but for this to work you will have to write separate names for same function with a different data-type and write a rust wrapper to handle these different data-types.
directly you can't do it ig cxx don't support generic vectors

Comment (SAtacker) [reply hpx-sys/src/lib.rs]: Got it, for now let's focus on arithmetic types like I32 only.
Thanks!

File: hpx-sys/include/wrapper.h [hunk 2]
```diff
@@ -44,4 +34,79 @@ inline std::int32_t disconnect_with_timeout(double shutdown_timeout, double loca
 
 inline std::int32_t finalize() { return hpx::finalize(); }
 
-/*inline std::int32_t stop() { return hpx::stop(); }*/
+inline void hpx_copy(const rust::Vec<int32_t>& src, rust::Vec<int32_t>& dest) {
+    std::vector<int32_t> cpp_src(src.begin(), src.end());
+    std::vector<int32_t> cpp_dest(dest.size());
+
+    hpx::copy(hpx::execution::par, cpp_src.begin(), cpp_src.end(), cpp_dest.begin());
```
Comment (SAtacker) [inline hpx-sys/include/wrapper.h]: Please remind me why we cannot use `src.begin()` instead of `cpp_src.begin()` , thanks!

**Comment by SAtacker** [comment]
> @SAtacker my hpx::copy_n is wrong, currently for first argument(as per docs refer to `beginning of the sequence of elements`) it would consider starting of src vec. To start copying from some other location in the src vector should i:
> 
> * increase arguments and add a offset variable
> * or use slice?

I don't have a preference given my limited experience with rust. I would prefer efficiency i.e. instead of creating an expensive temporary allocated array with vec.begin()+offset as starting element I would prefer no-copy reference to the remaining part of the elements since it is a copy operation. Does that makes sense, or did I miss anything?

File: hpx-sys/include/wrapper.h (hunk 2 above)
Comment (pingu-73) [reply hpx-sys/include/wrapper.h]: we have two vectors `src` and `cpp_src`. cpp_src is the vector that has elements from src ranges that user has defined in rust i.e `src.begin() to src.end()`

Comment (SAtacker) [reply hpx-sys/include/wrapper.h]: Please use src instead

File: hpx-sys/include/wrapper.h [hunk 3]
```diff
@@ -44,4 +34,181 @@ inline std::int32_t disconnect_with_timeout(double shutdown_timeout, double loca
...
+
+    dest.clear();
+    dest.reserve(cpp_dest.size());
+    for (const auto& item : cpp_dest) {
+        dest.push_back(item);
+    }
```
Comment (SAtacker) [inline hpx-sys/include/wrapper.h]: Please use `dest` instead of cpp_dest

File: hpx-sys/include/wrapper.h [hunk 4]
```diff
@@ -44,4 +34,181 @@ inline std::int32_t disconnect_with_timeout(double shutdown_timeout, double loca
...
+    std::vector<int32_t> cpp_dest(count);
+
+    hpx::copy_n(hpx::execution::par, cpp_src.begin(), count, cpp_dest.begin());
+
+    dest.clear();
+    dest.reserve(cpp_dest.size());
+    for (const auto& item : cpp_dest) {
+        dest.push_back(item);
+    }
+}
```
Comment (SAtacker) [inline hpx-sys/include/wrapper.h]: PTAL `hpx::copy`

**Review by SAtacker — CHANGES_REQUESTED** [review]
Please use the `src` vector instead of the `cpp_src` avoiding reconstruction.

**Comment by pingu-73** [comment]
@SAtacker this is the code-coverage report using [cargo-llvm-cov](https://github.com/taiki-e/cargo-llvm-cov)
### The following functions are not covered:
- finalize_with_timeout
- terminate
- disconnect
- disconnect_with_timeout
- merge
Screenshot 2024-08-29 at 9 10 47 PM

**Review by SAtacker — APPROVED** [review]

**Pull request merged**